src/java.base/share/classes/jdk/internal/math/FormattedFPDecimal.java

```diff
@@ -69,6 +68,74 @@ public static FormattedFPDecimal valueOf(double v, int prec, char form) {
         };
     }
 
+    private static FormattedFPDecimal split(double v) {
+        FormattedFPDecimal fd = new FormattedFPDecimal();
+        DoubleToDecimal.split(v, fd);
+        return fd;
+    }
+
+    /**
+     * Returns a FormattedFPDecimal with the appropriate precision for
+     * {@link Double#toString(double)}.
+     *
+     * @see java.math.BigDecimal#valueOf(double)
+     */
+    public static FormattedFPDecimal valueForDoubleToString(double v) {
+        final FormattedFPDecimal fd = split(v);
+        final int expR = fd.getExponentRounded();
+
+        // Adjust precision, following rules for Double.toString. There is
+        // always at least one digit and some cases require an extra one to
+        // force a digit after the decimal. No additional rounding is performed;
+        // no significant trailing digits are removed.
+
+        final int targetPrec =
+                // No extra trailing digit needed
+                (-3 <= expR && expR <= -1) ? 1
+
+                // Keep digits to left of decimal, plus leave a trailing zero
+                : (0 <= expR && expR <= 6) ? expR + 2 :
+
+                // Otherwise, require at least 2 digits, to include trailing
+                // digit when there is a single digit
+                2;
+
+
+        long s = fd.f;
+        int prec = fd.n;
+
+        if (prec < targetPrec) {
+            // Add zeros needed to reach target precision
+            int addZeros = targetPrec - prec;
```

Comment:
```suggestion
            final int addZeros = targetPrec - prec;
```
Just to be consistent with the usage of `final` in this method.